### PR TITLE
arch: cortex-m0: switch to asm!()

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request implements #1923 for cortex-m0.

### Testing Strategy

None? We don't have any cortex-m0 boards.

### TODO or Help Wanted

Help trying to verify my changes are OK.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: arch/cortex-m0/src/lib.rs [hunk 1]
```diff
@@ -162,47 +162,56 @@ pub unsafe extern "C" fn switch_to_user(
...
+    mrs r0, PSP /* PSP into user_stack */
 
-    mrs $0, PSP /* PSP into user_stack */
+    // Manually restore r6.
+    mov r6, r3
+
+    ",
+    inout("r0") user_stack,
+    in("r1") process_regs,
+    out("r3") _, out("r4") _, out("r5") _, out("r8") _, out("r9") _,
```
Comment (ppannuto) [inline arch/cortex-m0/src/lib.rs:212]: `r7`?

File: arch/cortex-m0/src/lib.rs [hunk 2]
```diff
@@ -286,19 +293,16 @@ _hardfault_msp:
...
 _hardfault_exit:
-    "
-    : "={r0}"(faulting_stack), "={r1}"(kernel_stack)
-    :
-    : "r0", "r1", "r2", "r3"
-    : "volatile"
-    );
+    ",
+    out("r0") faulting_stack, out("r1") kernel_stack,
+    out("r2") _, out("r3") _);
```
Comment (ppannuto) [inline arch/cortex-m0/src/lib.rs:298]: I believe this can add the `nomem` and `nostack` options

**Comment by bradjc** [comment]
Updated. Split hardfault handler to mimic cortex-m version since asm!() macro can be the only thing in a #[naked] function. Also save both r6 and r7 in context switch.

File: arch/cortex-m0/src/lib.rs [hunk 3]
```diff
@@ -111,7 +111,7 @@ MEXC_RETURN_MSP:
   .word 0xFFFFFFF9
 MEXC_RETURN_PSP:
   .word 0xFFFFFFFD"
-    : : : : "volatile");
+    );
```
Comment (hudson-ayers) [inline arch/cortex-m0/src/lib.rs:114]:
```suggestion
    , options(noreturn)
    );
```
the assembly manually jumps out. Also, naked functions require a single `noreturn` inline asm block -- it is UB for an inline asm block inside a naked function to return.

This suggestion will probably need formatting

File: arch/cortex-m0/src/lib.rs [hunk 4]
```diff
@@ -144,7 +144,7 @@ EXC_RETURN_MSP:
 EXC_RETURN_PSP:
   .word 0xFFFFFFFD
   "
-  : : : : "volatile"  );
+    );
```
Comment (hudson-ayers) [inline arch/cortex-m0/src/lib.rs:147]:
```suggestion
    , options(noreturn)
    );
```

**Review by hudson-ayers — CHANGES_REQUESTED** [review]

**Review by ppannuto — APPROVED** [review]

**Pull request merged**